protocol-units/bridge/move-modules/sources/atomic_bridge_initiator.move

```diff
@@ -0,0 +1,527 @@
+        store.nonce = store.nonce + 1;
+
+        // Create a single byte vector by concatenating all components
+        let combined_bytes = vector::empty<u8>();
+        vector::append(&mut combined_bytes, bcs::to_bytes(&addr));
+        vector::append(&mut combined_bytes, recipient);
+        vector::append(&mut combined_bytes, hash_lock);
+        vector::append(&mut combined_bytes, bcs::to_bytes(&store.nonce));
+
+        let bridge_transfer_id = aptos_hash::keccak256(combined_bytes);
+
+        let bridge_transfer = BridgeTransfer {
+            amount: amount,
+            originator: addr,
+            recipient: recipient,
+            hash_lock: hash_lock, 
+            time_lock: timestamp::now_seconds() + time_lock,
+            state: INITIALIZED,
+        };
+
+        vector::push_back(&mut store.transfers, bridge_transfer);
+
+        event::emit_event(&mut store.bridge_transfer_initiated_events, BridgeTransferInitiatedEvent {
+            bridge_transfer_id: bridge_transfer_id,
+            originator: addr,
+            recipient: recipient,
+            amount: amount,
+            hash_lock: hash_lock,
+            time_lock: timestamp::now_seconds() + time_lock,
+        });
+
+        bridge_transfer_id
+    }
+
+    public fun complete_bridge_transfer(
+        account: &signer,
+        bridge_transfer_id: vector<u8>,
+        pre_image: vector<u8>,
+        master_minter: &signer
+    ) acquires BridgeTransferStore, BridgeConfig {
+        let config_address = borrow_global<BridgeConfig>(@atomic_bridge).bridge_module_deployer;
+        let store = borrow_global_mut<BridgeTransferStore>(config_address);
+        let idx = get_bridge_transfer_index(&store.transfers, &bridge_transfer_id);
+        let bridge_transfer = vector::borrow_mut(&mut store.transfers, idx);
+
+        assert!(bridge_transfer.state == INITIALIZED, 1);
+        assert!(aptos_hash::keccak256(bcs::to_bytes(&pre_image)) == bridge_transfer.hash_lock, 2);
+        assert!(timestamp::now_seconds() <= bridge_transfer.time_lock, 3);
+
+        moveth::add_minter(master_minter, signer::address_of(account));
+
+        moveth::burn(account, @atomic_bridge, bridge_transfer.amount); 
+
+        bridge_transfer.state = COMPLETED;
+
+        event::emit_event(&mut store.bridge_transfer_completed_events, BridgeTransferCompletedEvent {
+            bridge_transfer_id: bridge_transfer_id,
+            pre_image: pre_image,
+        });
+    }
+
+    public fun refund_bridge_transfer(
+        account: &signer, 
+        bridge_transfer_id: vector<u8>,
+        atomic_bridge: &signer
+    ) acquires BridgeTransferStore, BridgeConfig {
+        let config_address = borrow_global<BridgeConfig>(@atomic_bridge).bridge_module_deployer;
+        let store = borrow_global_mut<BridgeTransferStore>(config_address);
+        let idx = get_bridge_transfer_index(&store.transfers, &bridge_transfer_id);
+        let bridge_transfer = vector::borrow_mut(&mut store.transfers, idx);
```

Comment:
Is this the only way to lookup the bridge_transfer, via its index?

Reply:
There are various `inline` helper functions in `[aptos_framework::vector](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/move-stdlib/sources/vector.move)` but they all ultimately borrow with ether `vector::borrow_mut` or `vector::borrow` using the index, so I don't see how it would improve performance to use those functions rather than a pattern like this. 

If this boilerplate is too bloated I could potentially refactor it into a helper function in this module?

Reply:
Up to you, if we need idx every time fine to leave it as it is .

Reply:
Guys, there must be another way, because this vector can become huge. We are not removing entries after they are completed, as we seem to want to keep them in state. This vector will accumulate to thousands of records. This would cost too much gas to process, especially since hashing is done on each visit. It's not even an option to keep it like this.

Reply:
Okay, I'll find a better way.

Reply:
I implemented smart_table to replace vector for the list of transfers, to eliminate the need for searching by index and so we could get rid of the extra helper function I wrote. 

Still need to handle removing transfers once complete, will do that next.